libkbfs/mdserver_memory.go

```diff
@@ -0,0 +1,641 @@
+	if err != nil {
+		return id, nil, MDServerError{err}
+	}
+	return id, nil, nil
+}
+
+// GetForTLF implements the MDServer interface for MDServerMemory.
+func (md *MDServerMemory) GetForTLF(ctx context.Context, id TlfID,
+	bid BranchID, mStatus MergeStatus) (*RootMetadataSigned, error) {
+	if md.isShutdown() {
+		return nil, errors.New("MD server already shut down")
+	}
+
+	if mStatus == Merged && bid != NullBranchID {
+		return nil, MDServerErrorBadRequest{Reason: "Invalid branch ID"}
+	}
+
+	mergedMasterHead, err :=
+		md.getHeadForTLF(ctx, id, NullBranchID, Merged)
+	if err != nil {
+		return nil, MDServerError{err}
+	}
+
+	// Check permissions
+	ok, err := isReader(
+		ctx, md.config.Codec(), md.config.KBPKI(), mergedMasterHead)
+	if err != nil {
+		return nil, MDServerError{err}
+	}
+	if !ok {
+		return nil, MDServerErrorUnauthorized{}
+	}
+
+	// Lookup the branch ID if not supplied
+	if mStatus == Unmerged && bid == NullBranchID {
+		bid, err = md.getBranchID(ctx, id)
+		if err != nil {
+			return nil, err
+		}
+		if bid == NullBranchID {
+			return nil, nil
+		}
+	}
+
+	rmds, err := md.getHeadForTLF(ctx, id, bid, mStatus)
+	if err != nil {
+		return nil, MDServerError{err}
+	}
+	return rmds, nil
+}
+
+func (md *MDServerMemory) getHeadForTLF(ctx context.Context, id TlfID,
+	bid BranchID, mStatus MergeStatus) (*RootMetadataSigned, error) {
+	key, err := md.getMDKey(id, bid, mStatus)
+	if err != nil {
+		return nil, err
+	}
+	md.mdMutex.Lock()
+	defer md.mdMutex.Unlock()
+	blocks := md.mdDb[key]
+	if len(blocks) == 0 {
+		return nil, nil
+	}
+	var rmds RootMetadataSigned
+	err = md.config.Codec().Decode(blocks[len(blocks)-1].encodedMd, &rmds)
+	if err != nil {
+		return nil, err
+	}
+	return &rmds, nil
+}
+
+func (md *MDServerMemory) getMDKey(
+	id TlfID, bid BranchID, mStatus MergeStatus) (mdBlockKey, error) {
+	if (mStatus == Merged) != (bid == NullBranchID) {
+		return mdBlockKey{},
+			fmt.Errorf("mstatus=%v is inconsistent with bid=%v",
+				mStatus, bid)
+	}
+	return mdBlockKey{id, bid}, nil
+}
+
+func (md *MDServerMemory) getBranchKey(ctx context.Context, id TlfID) (
+	mdBranchKey, error) {
+	// add device KID
+	deviceKID, err := md.getCurrentDeviceKID(ctx)
+	if err != nil {
+		return mdBranchKey{}, err
+	}
+	return mdBranchKey{id, deviceKID}, nil
+}
+
+func (md *MDServerMemory) getCurrentDeviceKID(ctx context.Context) (keybase1.KID, error) {
+	key, err := md.config.KBPKI().GetCurrentCryptPublicKey(ctx)
+	if err != nil {
+		return keybase1.KID(""), err
+	}
+	return key.kid, nil
+}
+
+// GetRange implements the MDServer interface for MDServerMemory.
+func (md *MDServerMemory) GetRange(ctx context.Context, id TlfID,
+	bid BranchID, mStatus MergeStatus, start, stop MetadataRevision) (
+	[]*RootMetadataSigned, error) {
+	md.log.CDebugf(ctx, "GetRange %d %d (%s)", start, stop, mStatus)
+	if md.isShutdown() {
+		return nil, errors.New("MD server already shut down")
+	}
+
+	if mStatus == Merged && bid != NullBranchID {
```

Comment:
Could factor out a good chunk of this function and share with `GetForTLF`, right?